Access token Request

eduardb opened the pull request:
We created two helper classes for the Twitter login flow. One that is a generic oauth helper, that creates a signature for a request and builds an authorization header. The second one is is for creating signed request to the Twitter API. For now we only have implemented the `requestToken` method that will provide a token for us to be used with login in the future.

We wrote tests for the `PercentEncoder` and `OauthHelper`, but we had to comment out the later ones because they were failing due to some react native stuff (CryptoJS to be more specific).

To test this functionality, we have included a debug screen accessible from the main screen (Oauth)

| iOS | Android |
| --- | --- |
| ![may-12-2016 17-36-15](https://cloud.githubusercontent.com/assets/2426348/15220518/10e51cf8-1868-11e6-94bd-34c384b9eeab.gif) | ![ezgif com-video-to-gif 1](https://cloud.githubusercontent.com/assets/2426348/15220505/02dee0bc-1868-11e6-9c6d-41cb17b2f088.gif) |

Paired with @gbasile

danybony commented on ReactNative/ReactTwitter/core/service/oauth-helper.js:
```diff
@@ -0,0 +1,75 @@
+    return CryptoJS.enc.Base64.stringify(HmacSHA1(signatureBase, signingKey))
+  }
+
+  static _getSignatureBase (method, url, params) {
```
Is this static for the tests?

eduardb replied:
To be honest, I wanted to avoid instantiating the class for the test, so yes 👅 
But either way, it's not using any object fields, so I don't see a reason why this shouldn't be static. Is it a bad practice?

danybony replied:
I don't know in Javascript, in general for OOP it's better to have instances of classes wit their specific methods

frapontillo replied:
I'm fairly new to ES6, or ES6 is fairly new to me, so I can't give a well formed opinion on this, but I'm good if we are going with the Java convention of making static any method that doesn't use any instance field. 👍

xrigau replied:
https://github.com/novoda/spikes/pull/35/files/2fcf0e570834fdf3ca86989f74b44a4b020f5ab3#r63572724

danybony commented on ReactNative/ReactTwitter/core/service/percent-encoder.js:
```diff
@@ -0,0 +1,7 @@
+class PercentEncoder {
```
What about `WebSafeEncoder`?

eduardb replied:
I used the terminology from the twitter dev docs here: https://dev.twitter.com/oauth/overview/percent-encoding-parameters

danybony commented on ReactNative/ReactTwitter/core/service/oauth-helper.js:
```diff
@@ -0,0 +1,75 @@
+    })
+  }
+
+  static _joinParams (params) {
```
Is it a Javascript convention to have a method used only in the current class starting with `_`?

eduardb replied:
No idea, I remember that I saw that somewhere, and that's why I used it this way. Maybe @frapontillo can tell us better?

frapontillo replied:
It's a convention to mark private methods with an underscore, yes.

xrigau replied:
also consider if it makes sense and if it makes it easier to read the code, don't just stick with it because it's a JS convention. Same way as we don't use mHungarian notation in android 🔍

eduardb replied:
There is no other way to tell if a method is private or not, and for me it's very readable...

xrigau replied:
👍

danybony commented on ReactNative/ReactTwitter/core/views/oauth-screen.js:
```diff
@@ -0,0 +1,63 @@
+        styleDisabled={styles.button_disabled}
+        onPress={this._buttonClicked}> Request Access Token </Button>
+        <Text style={styles.normal} numberOfLines={2}>
+          Access Token: {"\n"}{this.state.accessToken}
```
Do you need to pass the `\n` as a thing to be evaluated?

eduardb replied:
No, it's to break the line

xrigau commented:
👏

xrigau commented on ReactNative/ReactTwitter/__tests__/core/service/oauth-helper.test.js:
```diff
@@ -0,0 +1,41 @@
+
+describe('OauthHelper', () => {
+
+  // commented out because it's failing for a strange reason
```
?? why not just remove all this commented code?

eduardb replied:
Maybe someday someone will be able to figure out how to fix the tests 💀

frapontillo replied:
We can pair tomorrow if you need some help on this (should be pretty straightforward).

xrigau replied:
did this pairing happen?

eduardb replied:
Uhm, no. Maybe when Francesco comes back from I/O

frapontillo commented on ReactNative/ReactTwitter/__tests__/core/service/percent-encoder.test.js:
```diff
@@ -0,0 +1,17 @@
+
+describe('PercentEncoder', () => {
+  describe('encode', () => {
+    it('when encoding', () => {
```
You should really use the proper jasmine way of doing this. The pattern is:

```
it('should encode a special character', () => {
    // ...
})

it('should encode only special characters', () => {
    // ...
})

// etc
```

Don't write one `it` for multiple tests, or it'll break badly when (not if 👅) your code fails.

frapontillo commented on ReactNative/ReactTwitter/core/service/twitter-requests-service.js:
```diff
@@ -0,0 +1,51 @@
+var OauthHelper = require('../service/oauth-helper.js')
+
+const CONSUMER_KEY = 'aqPSTs1FT2ndP7qXi247BtbXd'
```
I'd load these from a `config.json`, it's easy as doing `const config = require('path/to/config.json')`

frapontillo commented on ReactNative/ReactTwitter/core/service/twitter-requests-service.js:
```diff
@@ -0,0 +1,51 @@
+      'oauth_version': '1.0'
+    }
+
+    return fetch(url, // eslint-disable-line no-undef
```
You shouldn't use globals, as they are unmockable. Wrap it in your own class if you need to.

frapontillo commented on ReactNative/ReactTwitter/core/service/twitter-requests-service.js:
```diff
@@ -0,0 +1,51 @@
+      })
+    .then((response) => { return response.text() })
+    .then((text) => { return this._getTokenDateFromResponse(text) })
+    .catch((error) => {
```
Or `catch(console.error)` 😉

frapontillo commented:
I left a bunch of comments, feel free to address anyone you agree with, accordingly with your time availability.

eduardb commented:
I addressed the comments. @frapontillo I don't feel that the tests that I commented out are a priority at this exact moment. Maybe I will get back to them as soon as we finish Twitter login. If you could help then figuring out what crashes, that would be great :)

frapontillo commented:
👍 for me, good work!

xrigau commented:
👍
